[mixed-content] Check if an URL is potentially trustworthy to block/upgrade

nicolo-ribaudo opened the pull request:
#### 3daacc928708d590bcb3dc6c7bb0c3ebf9992f8e
<pre>
[mixed-content] Check if an URL is potentially trustworthy to block/upgrade
<a href="https://bugs.webkit.org/show_bug.cgi?id=297536">https://bugs.webkit.org/show_bug.cgi?id=297536</a>

Reviewed by NOBODY (OOPS!).

The Mixed Content specification uses Secure Contexts&apos; &quot;potentially trustworthy&quot;
definition for URLs and origins to decide whether a request can be loaded as-is
or needs to be blocked/upgraded.

This is a larger bucket than just checking whether we have an HTTPs URL or not,
and includes embedder-defined custom protocols (such as, custom protocols
defined by WKWebView users). It matches the URLs for which
`window.isSecureContext` would return true.

This patch does not fully align WebKit&apos;s implementation with the Mixed Content
specification:
- handling of localhost URLs is left unchanged: they are always considered
  as potentially trustworthy, but we block HTTP loading of them in secure
  contexts (<a href="https://bugs.webkit.org/show_bug.cgi?id=171934)">https://bugs.webkit.org/show_bug.cgi?id=171934)</a>;
- loading of unsecure resources from file:// documents is still allowed
  (<a href="https://bugs.webkit.org/show_bug.cgi?id=297785).">https://bugs.webkit.org/show_bug.cgi?id=297785).</a>

These divergences are now explicitly marked and self-contained in the source
code, so that they can be easily changed if WebKit decides to align with
the spec.

* LayoutTests/imported/w3c/web-platform-tests/mixed-content/gen/top.http-rp/opt-in/sharedworker-import-data.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mixed-content/gen/top.meta/opt-in/sharedworker-import-data.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mixed-content/gen/top.meta/unset/sharedworker-import-data.https-expected.txt:
* Source/WebCore/fileapi/BlobURL.cpp:
(WebCore::BlobURL::getOwnerDocument):
(WebCore::BlobURL::isSecureBlobURL): Deleted.
* Source/WebCore/fileapi/BlobURL.h:
* Source/WebCore/loader/MixedContentChecker.cpp:
(WebCore::isNonLocalHostPotentiallyTrustworthyOrigin):
(WebCore::getActualOrigin):
(WebCore::isNonLocalHostPotentiallyTrustworthyURL):
(WebCore::prohibitsMixedSecurityContextsItself):
(WebCore::prohibitsMixedSecurityContexts):
(WebCore::MixedContentChecker::shouldUpgradeInsecureContent):
(WebCore::MixedContentChecker::canModifyRequest):
(WebCore::MixedContentChecker::shouldBlockRequest):
(WebCore::MixedContentChecker::checkFormForMixedContent):
(WebCore::isDocumentSecure): Deleted.
(WebCore::isDataContextSecure): Deleted.
(WebCore::isMixedContent): Deleted.
* Source/WebCore/page/SecurityOrigin.cpp:
(WebCore::SecurityOrigin::isSecure):
* Tools/TestWebKitAPI/Tests/WebCore/MixedContentChecker.cpp:
(TestWebKitAPI::TEST(MixedContentChecker, CanModifyRequest)):
</pre>


My original goal with this patch was to allow loading custom schemes from HTTPS pages in WKWebView, which should be allowed according to the Mixed Content spec. However, I noticed that the WebKit implementation of Mixed Content significantly diverges from the spec, which made that change complex. I ended up aligning the two with a couple exceptions where changing  the current WebKit behavior might have a too high impact.

- https://github.com/WebKit/WebKit/pull/35552 marked custom schemes as "upgradable" for compatibility with WKWebView ([bugs.webkit.org/show_bug.cgi?id=281883](https://bugs.webkit.org/show_bug.cgi?id=281883)). However, I don't understand how it can possibly work: once we decide that a custom scheme is upgradable, how do we actually upgrade it? It's not like HTTP that has a clear mapping to HTTPS. As far as I can tell we only upgrade http and ws requests ([356f67e/Source/WebCore/page/csp/ContentSecurityPolicy.cpp#L1151](https://github.com/WebKit/WebKit/blob/356f67eae83172534ddf23df2a69a017fbd05978/Source/WebCore/page/csp/ContentSecurityPolicy.cpp#L1151)).

- We have a setting to control upgrading of localhost (iPAddressAndLocalhostMixedContentUpgradeTestingEnabled), but in MixedContentChecker.cpp is only used for building error messages. Would it make sense to move its handling to here, rather than in the shouldUpgradeInsecureContent callers? (separate PR?)

- We have tome test (like LayoutTests/fast/forms/state-restore-form-associated-custom-elements-2.html and LayoutTests/fast/forms/form-associated-element-crash3.html) that we run with an origin whose protocol is file:, and that use non-trustworthy URLs (such as a `<form action="http:...">` in the first one, and a `<video src="javascript:...">`) in the second one. According to the Secure Contexts spec, file: URLs count as trustworthy so we should disallow mixed content in them. However, that breaks those tests and I don't know how to change the testing infrastructure to not serve them from file: but instead use http:. For now I'm preserving the old behavior of allowing insecure content to be used inside file:// and custom-scheme: documents ([bugs.webkit.org/show_bug.cgi?id=297785](https://bugs.webkit.org/show_bug.cgi?id=297785), https://github.com/w3c/webappsec-mixed-content/issues/73#issuecomment-3215055600).

- Right now I made sure to keep the spec divergence we have when it comes to http://127.0.0.1/ and http://localhost/ (they should both be considered secure, but we consider them upgradable instead - [bugs.webkit.org/show_bug.cgi?id=171934](https://bugs.webkit.org/show_bug.cgi?id=171934)). However, fixing it requires first changing how we run WPT (we use localhost in places where we shouldn't, I think?), which is way out of my knowledge level.

<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f380b39e951beb61ea1c13ae4edad6bd213c792b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120637 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40331 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30983 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127029 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72712 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122513 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41028 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48908 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91626 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60890 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123589 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32761 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108139 "Found 2 new API test failures: TestWebKitAPI.WebKit.ConfigurationMaskedURLSchemes_B, TestWebKitAPI.WebKit.ConfigurationMaskedURLSchemes (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72174 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31790 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26243 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70634 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102247 "Found 2 new API test failures: TestWebKitAPI.WebKit.ConfigurationMaskedURLSchemes_B, TestWebKitAPI.WebKit.ConfigurationMaskedURLSchemes (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26425 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129898 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47558 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36105 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100250 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47925 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104319 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100091 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45535 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23561 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/44218 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47420 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53125 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46889 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50235 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48575 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e8b1a14c2e675a91b148c1dbee2a0652a78b17a4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49713

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b593cb066b9976b44369a5dde3777d09194748d3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49713

nicolo-ribaudo commented:
(please don't review until I mark as ready! I want to get all the tests green first)

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/52d5d9164d3e7b355a061bbb53fa26124595494b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49713

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/cf94597bb7f3827b0cc4f8c32de95e04d6af0ef3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49713

nicolo-ribaudo committed:
[mixed-content] Check if an URL is potentially trustworthy to block/upgrade
https://bugs.webkit.org/show_bug.cgi?id=297536

Reviewed by NOBODY (OOPS!).

The Mixed Content specification uses Secure Contexts' "potentially trustworthy"
definition for URLs and origins to decide whether a request can be loaded as-is
or needs to be blocked/upgraded.

This is a larger bucket than just checking whether we have an HTTPs URL or not,
and includes embedder-defined custom protocols (such as, custom protocols
defined by WKWebView users). It matches the URLs for which
`window.isSecureContext` would return true.

This patch does not fully align WebKit's implementation with the Mixed Content
specification:
- handling of localhost URLs is left unchanged: they are always considered
  as potentially trustworthy, but we block HTTP loading of them in secure
  contexts (https://bugs.webkit.org/show_bug.cgi?id=171934);
- loading of unsecure resources from file:// documents is still allowed
  (https://bugs.webkit.org/show_bug.cgi?id=297785).

These divergences are now explicitly marked and self-contained in the source
code, so that they can be easily changed if WebKit decides to align with
the spec.

* LayoutTests/imported/w3c/web-platform-tests/mixed-content/gen/top.http-rp/opt-in/sharedworker-import-data.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mixed-content/gen/top.meta/opt-in/sharedworker-import-data.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mixed-content/gen/top.meta/unset/sharedworker-import-data.https-expected.txt:
* Source/WebCore/fileapi/BlobURL.cpp:
(WebCore::BlobURL::getOwnerDocument):
(WebCore::BlobURL::isSecureBlobURL): Deleted.
* Source/WebCore/fileapi/BlobURL.h:
* Source/WebCore/loader/MixedContentChecker.cpp:
(WebCore::isNonLocalHostPotentiallyTrustworthyOrigin):
(WebCore::getActualOrigin):
(WebCore::isNonLocalHostPotentiallyTrustworthyURL):
(WebCore::prohibitsMixedSecurityContextsItself):
(WebCore::prohibitsMixedSecurityContexts):
(WebCore::MixedContentChecker::shouldUpgradeInsecureContent):
(WebCore::MixedContentChecker::canModifyRequest):
(WebCore::MixedContentChecker::shouldBlockRequest):
(WebCore::MixedContentChecker::checkFormForMixedContent):
(WebCore::isDocumentSecure): Deleted.
(WebCore::isDataContextSecure): Deleted.
(WebCore::isMixedContent): Deleted.
* Source/WebCore/page/SecurityOrigin.cpp:
(WebCore::SecurityOrigin::isSecure):
* Tools/TestWebKitAPI/Tests/WebCore/MixedContentChecker.cpp:
(TestWebKitAPI::TEST(MixedContentChecker, CanModifyRequest)):

nicolo-ribaudo committed:
debug: split ConfigurationMaskedURLSchemes in many tests

nicolo-ribaudo committed:
debug: split ConfigurationMaskedURLSchemes further

nicolo-ribaudo committed:
debug: try handling the scheme

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5a2906c4f403afa5fcd36f2590f523caafecfb53. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49713

nicolo-ribaudo committed:
debug: also handle http

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0e896e235fcf3e3194048f50c3d534d43ee1f134. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49713

nicolo-ribaudo committed:
debug: More granular attempts

nicolo-ribaudo committed:
debug: split up synchronouslyLoadHTMLString

nicolo-ribaudo committed:
debug: timeout on waiting
